### PR TITLE
Switch feature detection to use xarray internally

### PR DESCRIPTION
**Pull request by freemansw1**
This is a draft PR that contains my work switching feature detection to use xarray internally, using many of the concepts (and some of the code) from #275. This code uses entirely xarray and pandas for processing feature detection, and removes all iris calls; instead using our internal conversion utilities to convert input iris data to xarray (which should always work, unlike `xarray`->`iris`. Further, this change should allow users to provide their data as an `xarray.DataArray`. 

This PR will remain in draft form until at least #293 is merged and intentionally includes those changes. 

More tests (particularly around times on different OS versions) should be added, but I wanted to get this into a form that we could discuss. 

* [x] Have you followed our guidelines in CONTRIBUTING.md? 
* [x] Have you self-reviewed your code and corrected any misspellings? 
* [x] Have you written documentation that is easy to understand?
* [x] Have you written descriptive commit messages? 
* [x] Have you added NumPy docstrings for newly added functions? 
* [x] Have you formatted your code using black? 
* [x] If you have introduced a new functionality, have you added adequate unit tests?
* [x] Have all tests passed in your local clone? 
* [ ] If you have introduced a new functionality, have you added an example notebook?
* [ ] Have you kept your pull request small and limited so that it is easy to review? 
* [x] Have the newest changes from this branch been merged?

### CODE REVIEW

**Comment by freemansw1** [comment]
Okay, this is now ready for review in my opinion. This PR contains changes from #380.

I'd like feedback (and testing) everywhere, but in particular, I'd like feedback on the new `add_coordinates_to_features` function in `tobac/utils/internal/xarray_utils.py`. I had to come up with a clever solution, and I really hate clever solutions.

**Comment by w-k-jones** [comment]
> Okay, this is now ready for review in my opinion. This PR contains changes from #380.
> 
> I'd like feedback (and testing) everywhere, but in particular, I'd like feedback on the new `add_coordinates_to_features` function in `tobac/utils/internal/xarray_utils.py`. I had to come up with a clever solution, and I really hate clever solutions.

Aha, multidimensional interpolation with xarray is a massive pain...

**Comment by freemansw1** [comment]
> > Okay, this is now ready for review in my opinion. This PR contains changes from #380.
> > I'd like feedback (and testing) everywhere, but in particular, I'd like feedback on the new `add_coordinates_to_features` function in `tobac/utils/internal/xarray_utils.py`. I had to come up with a clever solution, and I really hate clever solutions.
> 
> Aha, multidimensional interpolation with xarray is a massive pain...

The issue actually came about because we have `i,j,k` coordinates, and there doesn't appear to be a way to interpolate along those coordinates if a coord and dim share a name. The multidimensional aspect actually wasn't too bad as it only has to interpolate a single point at a time.

**Comment by w-k-jones** [comment]
> > > Okay, this is now ready for review in my opinion. This PR contains changes from #380.
> > > I'd like feedback (and testing) everywhere, but in particular, I'd like feedback on the new `add_coordinates_to_features` function in `tobac/utils/internal/xarray_utils.py`. I had to come up with a clever solution, and I really hate clever solutions.
> > 
> > 
> > Aha, multidimensional interpolation with xarray is a massive pain...
> 
> The issue actually came about because we have `i,j,k` coordinates, and there doesn't appear to be a way to interpolate along those coordinates if a coord and dim share a name. The multidimensional aspect actually wasn't too bad as it only has to interpolate a single point at a time.

Ah I see what you mean, it's because of interpolating based on array index location rather than coord value

**Comment by w-k-jones** [comment]
Could this be solved using [DataArray.swap_dims](https://docs.xarray.dev/en/latest/generated/xarray.DataArray.swap_dims.html)?:
```
variable_da = variable_da.swap_dims({"time":"frame", "south_north":"hdim_1", "west_east":"hdim_2"})
variable_da.interp(
    coords=dict(
        frame=xr.DataArray(Features["frame"].values, dims="feature"),
        hdim_1=xr.DataArray(Features["hdim_1"].values, dims="feature"),
        hdim_2=xr.DataArray(Features["hdim_2"].values, dims="feature"),
    )
)
```

Suggestion was found via this still open xarray issue from 4 years ago... https://github.com/pydata/xarray/issues/3343

**Comment by freemansw1** [comment]
> Could this be solved using [DataArray.swap_dims](https://docs.xarray.dev/en/latest/generated/xarray.DataArray.swap_dims.html)?:
> 
> ```
> variable_da = variable_da.swap_dims({"time":"frame", "south_north":"hdim_1", "west_east":"hdim_2"})
> variable_da.interp(
>     coords=dict(
>         frame=xr.DataArray(Features["frame"].values, dims="feature"),
>         hdim_1=xr.DataArray(Features["hdim_1"].values, dims="feature"),
>         hdim_2=xr.DataArray(Features["hdim_2"].values, dims="feature"),
>     )
> )
> ```
> 
> Suggestion was found via this still open xarray issue from 4 years ago... [pydata/xarray#3343](https://github.com/pydata/xarray/issues/3343)

Yes, good catch. I hadn't seen that; I came up with the same pathway, but was doing a `DataArray`->`Dataset` conversion first because I thought it would be named something sensible like `rename_dims`. I've revised this now.

File: tobac/utils/internal/xarray_utils.py
Comment (w-k-jones) [inline tobac/utils/internal/xarray_utils.py:382]: I'm not a fan of using randomly generated names here. I would rather use a set of fixed names, similar to what we use for renaming dataframe columns for predicitive tracking. e.g. `__temp_hdim1_name`, `__temp_vdim_name`

File: tobac/utils/internal/xarray_utils.py
Comment (w-k-jones) [inline tobac/utils/internal/xarray_utils.py:390]: Suggestion: use `drop_vars` instead of the depreciated `drop` here, and use `errors=ignore` to avoid the if statement:

```suggestion
    interpolated_df = interpolated_df.drop_vars([hdim1_name_new, hdim2_name_new, vdim_name_new], errors="ignore")
```

**Comment by freemansw1** [comment]
Note to self: this depends on #380 to merge.

File: tobac/feature_detection.py
Comment (freemansw1) [reply tobac/feature_detection.py:1245]: I have deliberately not done this for now as we don't really have support for it throughout. It would take a pretty substantial refactor (I think) to ensure support all the way through.

File: tobac/utils/internal/xarray_utils.py
Comment (freemansw1) [reply tobac/utils/internal/xarray_utils.py:13]: I've removed the unused imports here.

File: tobac/utils/internal/xarray_utils.py
Comment (freemansw1) [reply tobac/utils/internal/xarray_utils.py:61]: Agreed; I've switched this to return an exception rather than `None`.

File: tobac/utils/internal/xarray_utils.py
Comment (freemansw1) [reply tobac/utils/internal/xarray_utils.py:40]: Resolved

File: tobac/feature_detection.py
Comment (freemansw1) [reply tobac/feature_detection.py:1253]: Agreed, I've done this now.

File: tobac/feature_detection.py
Comment (freemansw1) [reply tobac/feature_detection.py:1345]: Agreed. `isel` seems to have reasonable performance, too.

File: tobac/utils/internal/basic.py
Comment (freemansw1) [reply tobac/utils/internal/basic.py:119]: Good idea. Done.

**Comment by freemansw1** [comment]
Got about halfway through the review comments... Next few will take a bit longer.

File: tobac/utils/internal/xarray_utils.py
Comment (freemansw1) [reply tobac/utils/internal/xarray_utils.py:99]: Fair point, resolved.

File: tobac/utils/internal/xarray_utils.py
Comment (freemansw1) [reply tobac/utils/internal/xarray_utils.py:138]: Done.

File: tobac/utils/internal/xarray_utils.py
Comment (freemansw1) [reply tobac/utils/internal/xarray_utils.py:149]: Agreed. Kept it simple for now.

File: tobac/utils/internal/xarray_utils.py
Comment (freemansw1) [reply tobac/utils/internal/xarray_utils.py:380]: Yeah, adding in the random names is not a great idea. I've taken your suggestion here and added a check to see if these names are otherwise used, these are reasonable barriers to add.

File: tobac/utils/internal/xarray_utils.py
Comment (freemansw1) [reply tobac/utils/internal/xarray_utils.py:388]: Agreed. I've refactored this.

**Comment by freemansw1** [comment]
OK, I have resolved all review comments here. @w-k-jones ready for a second review. We will need another review from someone else, too, but this is not urgent as I need to get #380 resolved first.

**Review by w-k-jones — APPROVED** [review]
Great, thanks for making the revisions. I've had a look through the changes and tested the code with both xarray and iris locally and have no issues to issues. Should we start an `RC_v1.6.0` branch to merge this into?

**Comment by freemansw1** [comment]
> Great, thanks for making the revisions. I've had a look through the changes and tested the code with both xarray and iris locally and have no issues to issues. Should we start an `RC_v1.6.0` branch to merge this into?

Yes, we will need to do that. I'll wait until we have to though to minimize the amount of time we have to maintain separate branches. We still need one more reviewer, @JuliaKukulies or @kelcyno are you up for it?

**Comment by JuliaKukulies** [comment]
> > Great, thanks for making the revisions. I've had a look through the changes and tested the code with both xarray and iris locally and have no issues to issues. Should we start an `RC_v1.6.0` branch to merge this into?
> 
> Yes, we will need to do that. I'll wait until we have to though to minimize the amount of time we have to maintain separate branches. We still need one more reviewer, @JuliaKukulies or @kelcyno are you up for it?

I can do the second review within the next couple of days.

File: tobac/utils/bulk_statistics.py
Comment (JuliaKukulies) [inline tobac/utils/bulk_statistics.py:27]: is it not better to keep the labels as integers?

**Review by JuliaKukulies — COMMENTED** [review]
Thanks a lot for this pR @freemansw1! This is seems like a great step towards the xarray transition.

It all looks good to me besides two small comments. And I have tested this on two real data sets which works fine and without any major issues. However, the only thing that does not seem to work is is passing the argument `preserve_iris_datetime_type` to `feature_detection_multithreshold()`. The dtypes for time are different in the output feature dataframe no matter if I set `preserve_iris_datetime_type` to False or True. They are always `objects/dtype('O')` when I input an iris cube and `datetime64/dtype('<M8[ns]')` when I input an xarray.

**Comment by freemansw1** [comment]
Ok, @JuliaKukulies, I think I've addressed your concerns. Ready for your re-review. Hopefully we can get this merged!

**Comment by JuliaKukulies** [comment]
Excellent, @freemansw1! This looks good except for that there is still the  issue with `preserve_iris_datetime_type` which does not result in the same datetime format as when I provide iris input when this parameters is `True`  (see my last comment). 

Sorry I had not marked the bug at the actual code locations, so I get that you might have overseen this. I also cannot really figure out where this bug comes from, since everywhere in the code where `preserve_iris_datetime_type` is used looks like it should work. Could you reproduce this or is this an issue with my test data?

**Comment by freemansw1** [comment]
> Excellent, @freemansw1! This looks good except for that there is still the issue with `preserve_iris_datetime_type` which does not result in the same datetime format as when I provide iris input when this parameters is `True` (see my last comment).
> 
> Sorry I had not marked the bug at the actual code locations, so I get that you might have overseen this. I also cannot really figure out where this bug comes from, since everywhere in the code where `preserve_iris_datetime_type` is used looks like it should work. Could you reproduce this or is this an issue with my test data?

Apologies, yeah, I had missed this. I will work to reproduce and try this out this weekend/early next week.

**Comment by w-k-jones** [comment]
I think I've noticed this before, and the issue is that for a `_cftime.DatetimeProlepticGregorian` a `_cftime.datetime` object is returned. This doesn't seem to cause any issues but could cause problems downstream... I think to address it we would have to perform the conversion in the decorator as we need to know the dtype of the input iris cube to preserve the exact cftime type

**Comment by freemansw1** [comment]
Note to self: `add_coords` for xarray fails when having coordinates to interpolate along dimensions other than x/y.

**Comment by freemansw1** [comment]
OK, I *think* I have resolved the datetime issues. It's somewhat of a hacky solution, but basically, `xarray` only converts to `datetime64` if the time received is in Gregorian time. I now check for this and convert back to Gregorian time. For all other calendars (thanks @harrietgilmour for helping with this part), xarray leaves it as an object (which does still break our code, but alas).

**Comment by freemansw1** [comment]
> Note to self: `add_coords` for xarray fails when having coordinates to interpolate along dimensions other than x/y.

I have now also resolved this issue. The solution here results in some decisions that were contrary to the discussion we had at the last dev meeting (notably that len(1) coordinates are still appended to the dataframe), but I think we are otherwise okay. We can talk about this at the dev meeting tomorrow.

**Comment by freemansw1** [comment]
@w-k-jones and @JuliaKukulies I am now ready for a re-review.

**Comment by JuliaKukulies** [comment]
> OK, I _think_ I have resolved the datetime issues. It's somewhat of a hacky solution, but basically, `xarray` only converts to `datetime64` if the time received is in Gregorian time. I now check for this and convert back to Gregorian time. For all other calendars (thanks @harrietgilmour for helping with this part), xarray leaves it as an object (which does still break our code, but alas).

OK, I have tested this again with my data and I thought I understood what you explained in the dev meeting today, but now i am a little confused again :) I basically get: 

- for xarray input and `preserve_iris_datetime_type = True` -> `datetime64`
- for xarray input and `preserve_iris_datetime_type = False` -> `datetime64`

- for iris input and `preserve_iris_datetime_type = True` -> `object (cftime.DatetimeGregorian)`
- for iris input and `preserve_iris_datetime_type = False` ->  `object (cftime.DatetimeGregorian)` 

Is that what we want?

**Comment by freemansw1** [comment]
> > OK, I _think_ I have resolved the datetime issues. It's somewhat of a hacky solution, but basically, `xarray` only converts to `datetime64` if the time received is in Gregorian time. I now check for this and convert back to Gregorian time. For all other calendars (thanks @harrietgilmour for helping with this part), xarray leaves it as an object (which does still break our code, but alas).
> 
> OK, I have tested this again with my data and I thought I understood what you explained in the dev meeting today, but now i am a little confused again :) I basically get:
> 
> * for xarray input and `preserve_iris_datetime_type = True` -> `datetime64`
> * for xarray input and `preserve_iris_datetime_type = False` -> `datetime64`
> * for iris input and `preserve_iris_datetime_type = True` -> `object (cftime.DatetimeGregorian)`
> * for iris input and `preserve_iris_datetime_type = False` ->  `object (cftime.DatetimeGregorian)`
> 
> Is that what we want?

Ok, the first of those three are all correct, but the last one isn't.

**Comment by JuliaKukulies** [comment]
> > > OK, I _think_ I have resolved the datetime issues. It's somewhat of a hacky solution, but basically, `xarray` only converts to `datetime64` if the time received is in Gregorian time. I now check for this and convert back to Gregorian time. For all other calendars (thanks @harrietgilmour for helping with this part), xarray leaves it as an object (which does still break our code, but alas).
> > 
> > 
> > OK, I have tested this again with my data and I thought I understood what you explained in the dev meeting today, but now i am a little confused again :) I basically get:
> > 
> > * for xarray input and `preserve_iris_datetime_type = True` -> `datetime64`
> > * for xarray input and `preserve_iris_datetime_type = False` -> `datetime64`
> > * for iris input and `preserve_iris_datetime_type = True` -> `object (cftime.DatetimeGregorian)`
> > * for iris input and `preserve_iris_datetime_type = False` ->  `object (cftime.DatetimeGregorian)`
> > 
> > Is that what we want?
> 
> Ok, the first of those three are all correct, but the last one isn't.

Thats what I thought, and we would expect `datetime64` for the last one, right? Have you tested this with the data I sent you or another dataset? Sorry to cause more struggle..

**Comment by freemansw1** [comment]
> > > > OK, I _think_ I have resolved the datetime issues. It's somewhat of a hacky solution, but basically, `xarray` only converts to `datetime64` if the time received is in Gregorian time. I now check for this and convert back to Gregorian time. For all other calendars (thanks @harrietgilmour for helping with this part), xarray leaves it as an object (which does still break our code, but alas).
> > > 
> > > 
> > > OK, I have tested this again with my data and I thought I understood what you explained in the dev meeting today, but now i am a little confused again :) I basically get:
> > > 
> > > * for xarray input and `preserve_iris_datetime_type = True` -> `datetime64`
> > > * for xarray input and `preserve_iris_datetime_type = False` -> `datetime64`
> > > * for iris input and `preserve_iris_datetime_type = True` -> `object (cftime.DatetimeGregorian)`
> > > * for iris input and `preserve_iris_datetime_type = False` ->  `object (cftime.DatetimeGregorian)`
> > > 
> > > Is that what we want?
> > 
> > 
> > Ok, the first of those three are all correct, but the last one isn't.
> 
> Thats what I thought, and we would expect `datetime64` for the last one, right? Have you tested this with the data I sent you or another dataset? Sorry to cause more struggle..

Ok this one was easy- we didn't actually expect the kwarg in `feature_detection_multithreshold`. I've fixed that.

**Comment by JuliaKukulies** [comment]
> > > > > OK, I _think_ I have resolved the datetime issues. It's somewhat of a hacky solution, but basically, `xarray` only converts to `datetime64` if the time received is in Gregorian time. I now check for this and convert back to Gregorian time. For all other calendars (thanks @harrietgilmour for helping with this part), xarray leaves it as an object (which does still break our code, but alas).
> > > > 
> > > > 
> > > > OK, I have tested this again with my data and I thought I understood what you explained in the dev meeting today, but now i am a little confused again :) I basically get:
> > > > 
> > > > * for xarray input and `preserve_iris_datetime_type = True` -> `datetime64`
> > > > * for xarray input and `preserve_iris_datetime_type = False` -> `datetime64`
> > > > * for iris input and `preserve_iris_datetime_type = True` -> `object (cftime.DatetimeGregorian)`
> > > > * for iris input and `preserve_iris_datetime_type = False` ->  `object (cftime.DatetimeGregorian)`
> > > > 
> > > > Is that what we want?
> > > 
> > > 
> > > Ok, the first of those three are all correct, but the last one isn't.
> > 
> > 
> > Thats what I thought, and we would expect `datetime64` for the last one, right? Have you tested this with the data I sent you or another dataset? Sorry to cause more struggle..
> 
> Ok this one was easy- we didn't actually expect the kwarg in `feature_detection_multithreshold`. I've fixed that.

Perfect, thanks for looking into this and fixing this. I am happy to approve this now once the latest changes from RC_v1.5.x are merged in

**Comment by freemansw1** [comment]
OK, changes all merged in and everything. @JuliaKukulies and @w-k-jones are you happy for me to merge into `RC_v1.6.0`?

**Comment by JuliaKukulies** [comment]
> OK, changes all merged in and everything. @JuliaKukulies and @w-k-jones are you happy for me to merge into `RC_v1.6.0`?

Yes, please go ahead:) !

**Comment by w-k-jones** [comment]
Also happy for this to be merged :)

**Pull request merged**
